src/exe.ml

```diff
@@ -138,6 +143,72 @@ let link_exe
       Build.dyn_paths (Build.arr (fun (modules, _) ->
         artifacts modules ~ext:ctx.ext_obj))
   in
+  let link_flags_for_requires =
+    Result.map requires ~f:(fun libs ->
+      match cut_after [] "findlib.dynload" libs with
+      | Some (before, after) ->
+        (* If findlib.dynload is linked, we stores in the binary the packages linked by linking just
+           after findlib.dynload a module containing the info *)
+        let libs =
+          List.filter
+            ~f:(fun lib -> match Lib.status lib with | Lib.Status.Private _ -> false | _ -> true)
+            libs
+        in
+        (* We can't use directly Findlib.Package.preds because there
+           are fixups in ocamlfind that adds the package "thread" as
+           dependency of all the packages when the predicate "mt" is
+           used. Also independently of ocamlfind it could be strange
+           if a library can be used without or with `thread` but
+           "thread" is forced during dynlink because we always add "mt". *)
+        let preds = if List.exists libs ~f:(fun lib -> Lib.name lib = "threads")
+          then
+            Findlib.Package.preds
+          else
+            Variant.Set.diff Findlib.Package.preds (Variant.Set.of_list [Variant.mt;Variant.mt_posix])
+        in
+        let preds = Variant.Set.add preds (Mode.variant mode) in
+        let s =
+          Format.asprintf "%a@\nFindlib.record_package_predicates %a;;@."
+            (Fmt.list ~pp_sep:Fmt.nl (fun fmt lib -> Format.fprintf fmt "Findlib.record_package Findlib.Record_core %S;;" (Lib.name lib)))
+            libs
+            (Fmt.ocaml_list Variant.pp) (Variant.Set.to_list preds)
+        in
+        let basename = Format.asprintf "%s_findlib_initl_%a" name Mode.pp mode in
+        let ml  = Path.relative dir (basename ^ ".ml") in
+        let mli = Path.relative dir (basename ^ ".mli") in
```

Comment:
In general we try to put such generated modules outside of the source directory. Could you put it in the object directory instead? This is to avoid it being included in `(glob_files *.ml)`.

Additionally, we don't generate a mli for other such generated modules (utop, inline tests, ...) so for consistency we shouldn't do it here either.